compiler/rustc_errors/src/json.rs

```diff
@@ -135,7 +135,12 @@ impl Emitter for JsonEmitter {
         let data: Vec<FutureBreakageItem<'_>> = diags
             .into_iter()
             .map(|mut diag| {
-                if diag.level == crate::Level::Allow {
+                // The `FutureBreakageItem` is collected and serialized.
+                // However, the `allow` and `expect` lint levels can't usually
+                // be serialized. The lint level is overwritten to allow the
+                // serialization again and force a lint emission.
+                // (This is an educated guess. I didn't originally add this)
+                if matches!(diag.level, crate::Level::Allow | crate::Level::Expect(..)) {
```

Comment:
I think this is mostly right. I looked at the introduction of [those lines](https://github.com/rust-lang/rust/commit/23018a55d9735afbefb19fcec91db4b53fe917d8#diff-4709ec799c1d3ac3ab848fc853bcf0250fae27935fe67a394a4832da9d7fa9b2R139-R141) and they come straight from the introduction of future incompatibility system.

Interestingly the PR also introduced the `allow` level, which I think is because before that PR an allowed lint could never trigger a lint, since well it's allowed, but future incompatibility forces the lint emission even through it's current level is `allow`, and the same happens with `expect`.

And because the emitter system doesn't expect `allow` or `expect` as they don't normally make sense, we change (just before emiting the lint) it's level to `warn` as to make the emitter happy but also to avoid confused users.

I would therefore suggest changing the text to something like this:
```suggestion
                // Allowed or expected lints don't normally (by definition) emit a lint
                // but future incompat lints are special and are emitted anyway.
                //
                // So to avoid ICEs and confused users we "upgrade" the lint level for
                // those `FutureBreakageItem` to warn.
                if matches!(diag.level, crate::Level::Allow | crate::Level::Expect(..)) {
```

Reply:
This is amazing. Thank you for tracking down the actual source and suggesting this. :heart:

Reply:
This suggestion hasn't been applied up until r+. @Urgau are you okay with this? Is this suggestion optional?

Reply:
Um... no, I expect this suggestion to be applied.
Let me do it from the GitHub UI.